### PR TITLE
jira handshake

### PR DESCRIPTION
**Pull request by lwsanty**
close [[DEV-859]]

Signed-off-by: lwsanty <lwsanty@gmail.com>

[DEV-859]: https://athenianco.atlassian.net/browse/DEV-859

### CODE REVIEW

File: jira/client/transport/jwt.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+func qsh(r *http.Request) string {
+	return r.Method + "&" + r.URL.EscapedPath() + "&" + qshValues(r)
+}
+
+type qshVar struct {
+	Key string
+	Val string
+}
+
+func (v qshVar) string() string {
```
Comment (dennwc) [inline jira/client/transport/jwt.go:80]:
```suggestion
func (v qshVar) String() string {
```

File: jira/client/transport/jwt.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	}), nil
+}
+
+func sha(s string) (string, error) {
+	hash := sha256.New()
+	if _, err := io.Copy(hash, strings.NewReader(s)); err != nil {
+		return "", err
+	}
+	sum := hash.Sum(nil)
+	return fmt.Sprintf("%x", sum), nil
```
Comment (dennwc) [inline jira/client/transport/jwt.go:129]: `hex.EncodeToString`

File: jira/client/transport/jwt.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		"qsh": qshEncoded,
+		"iat": now.Unix(),
+		"exp": now.Add(tokenTimeout).Unix(),
+		"iss": key,
+	}), nil
+}
+
+func sha(s string) (string, error) {
+	hash := sha256.New()
+	if _, err := io.Copy(hash, strings.NewReader(s)); err != nil {
```
Comment (dennwc) [inline jira/client/transport/jwt.go:125]:
```
hash.Write([]byte(s))
```

File: jira/client/transport/unified.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+func NewUnifiedAuthTransport() *UnifiedAuthTransport {
+	return &UnifiedAuthTransport{
+		basic: &BasicAuthTransport{},
+		jwt:   &JWTAuthTransport{},
+	}
+}
+
+func (t *UnifiedAuthTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	ctx := req.Context()
+	req2 := cloneRequest(req)
```
Comment (dennwc) [inline jira/client/transport/unified.go:24]: It should be safe to pass the original `req` here. No need to clone.

File: jira/client/transport/unified.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	_, ok := types.UserContext(ctx)
+	if ok {
+		return t.basic.RoundTrip(req2)
+	}
+	_, ok = types.InstallationContext(ctx)
+	if ok {
+		return t.jwt.RoundTrip(req2)
+	}
+	return nil, fmt.Errorf("failed to route the request")
```
Comment (dennwc) [inline jira/client/transport/unified.go:33]: `"no auth credentials set"` would be a better explanation; also, `errors.New`.

File: jira/database/database.go [hunk 6]
```diff
@@ -18,12 +18,20 @@ type ProjectID = types.ProjectID
 type ProjectKey = types.ProjectKey
 type Component = types.Component
 type ComponentID = types.ComponentID
+type Installation = types.Installation
 
 // TODO: CRUD account
 type Database interface {
 	// TODO: iterator? temporal solution
 	GetAccIDs(ctx context.Context) ([]types.AccID, error)
```
Comment (dennwc) [inline jira/database/database.go:26]: Is it still used?

File: jira/database/database_pg.go [hunk 7]
```diff
@@ -573,6 +573,85 @@ func (db *pgDatabase) ListComponents(ctx context.Context, accID AccID, projectID
 	return &componentIter{rows: rows}, nil
 }
 
+const installationColumns = "acc_id, key, server_version, plugins_version, base_url, product_type, description"
+
+// TODO: maybe restrict overwrites and force uninstall(delete)
+func (db *pgDatabase) CreateInstallation(ctx context.Context, i *Installation) (AccID, error) {
+	var id int64
+	if err := db.QueryRowContext(ctx, `INSERT INTO
```
Comment (dennwc) [inline jira/database/database_pg.go:581]: It's really hard to read. Can you indent it the same way as other queries (for example in the GH package)?

File: sql/migrations/015_jira_installation.up.sql [hunk 8]
```diff
@@ -0,0 +1,11 @@
+-- TODO: be more strict and add constraints for jira server detection
+CREATE TABLE jira.installation (
+    acc_id BIGSERIAL NOT NULL,
```
Comment (dennwc) [inline sql/migrations/015_jira_installation.up.sql:3]: This must be a `PRIMARY KEY`.

File: sql/migrations/015_jira_installation.up.sql [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+    key text,
+    server_version text,
+    plugins_version text,
+    base_url text NOT NULL,
```
Comment (dennwc) [inline sql/migrations/015_jira_installation.up.sql:7]: Instead of making this a PK, define it as `UNIQUE`.

File: pkeyprocessor/pkeyprocessor.go [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package pkeyprocessor
+
+import "context"
+
+// TODO: move to https://github.com/athenianco/cloud-common
```
Comment (dennwc) [inline pkeyprocessor/pkeyprocessor.go:5]: Please do.

File: jira/database/database.go [hunk 11]
```diff
@@ -18,12 +18,20 @@ type ProjectID = types.ProjectID
...
 
+	// installations
+	CreateInstallation(ctx context.Context, i *Installation) (AccID, error)
+	GetInstallation(ctx context.Context, accID AccID) (*Installation, error)
+	GetInstallationByURL(ctx context.Context, url string) (*Installation, error)
```
Comment (dennwc) [inline jira/database/database.go:31]: Seems unused.

**Review by dennwc — CHANGES_REQUESTED** [review]

File: athenian/database_pg.go [hunk 12]
```diff
@@ -52,7 +52,18 @@ func scanAccount(sc dbs.Scanner) (Account, error) {
 }
 
 func (db *database) GetAccount(ctx context.Context, id AccountID) (*Account, error) {
-	row := db.db.QueryRow(ctx, `SELECT id, create_at, secret, secret_salt FROM public.accounts WHERE id = $1`, int64(id))
+	row := db.db.QueryRow(ctx, `SELECT id, created_at, secret, secret_salt FROM public.accounts WHERE id = $1`, int64(id))
```
[the remark this answers is not in the document]
Comment (lwsanty) [reply athenian/database_pg.go]: Done.

File: jira/types/types.go [hunk 13]
```diff
@@ -75,6 +75,38 @@ type Component struct {
 	Name      string
 }
 
+// Installation is used for DB
+type Installation struct {
+	AccID          AccID
```
[the remark this answers is not in the document]
Comment (lwsanty) [reply jira/types/types.go]: Done.

File: jira_webhook.go [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+	"strings"
+
+	"github.com/athenianco/cloud-common/report"
+	"github.com/athenianco/metadata/athenian"
+	"github.com/athenianco/metadata/jira/database"
+	"github.com/julienschmidt/httprouter"
+)
+
+const (
+	helloWorld = `<!DOCTYPE html>
```
[the remark this answers is not in the document]
Comment (lwsanty) [reply jira_webhook.go]: Done.

File: jira_webhook.go [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+    "url": "https://athenian.co"
+  },
+  "links": {
+    "self": "%[2]s/atlassian-connect.json"
+  },
+  "lifecycle": {
+    "installed": "/installed",
+    "uninstalled": "/uninstalled"
+  },
+  "baseUrl": "%[2]s",
```
[the remark this answers is not in the document]
Comment (lwsanty) [reply jira_webhook.go:40]: could you please elaborate?

File: jira/client/transport/jwt.go (hunk 1 above)
Comment (lwsanty) [reply jira/client/transport/jwt.go]: Done.

File: jira/database/database_pg.go (hunk 7 above)
Comment (lwsanty) [reply jira/database/database_pg.go]: Done.

File: sql/migrations/015_jira_installation.up.sql (hunk 8 above)
Comment (lwsanty) [reply sql/migrations/015_jira_installation.up.sql]: Done.

File: sql/migrations/015_jira_installation.up.sql (hunk 9 above)
Comment (lwsanty) [reply sql/migrations/015_jira_installation.up.sql]: Done.

File: pkeyprocessor/pkeyprocessor.go (hunk 10 above)
Comment (lwsanty) [reply pkeyprocessor/pkeyprocessor.go]: Done.

File: jira/database/database.go (hunk 11 above)
Comment (lwsanty) [reply jira/database/database.go]: Done.

File: jira_webhook.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (lwsanty) [reply jira_webhook.go]: Done.

File: jira/database/database.go (hunk 6 above)
Comment (lwsanty) [reply jira/database/database.go]: Done.

File: jira/client/transport/unified.go (hunk 5 above)
Comment (lwsanty) [reply jira/client/transport/unified.go]: Done.

File: jira/client/transport/jwt.go (hunk 3 above)
Comment (lwsanty) [reply jira/client/transport/jwt.go]: Done.

File: jira/client/transport/unified.go (hunk 4 above)
Comment (lwsanty) [reply jira/client/transport/unified.go]: Done.

File: jira/client/transport/jwt.go (hunk 2 above)
Comment (lwsanty) [reply jira/client/transport/jwt.go]: Done.

**Comment by dennwc** [comment]
Some conflicts here. Will soon review it again.

**Review by dennwc — COMMENTED** [review]
@lwsanty LGTM! Now the important question: have you tested it in action? :)

**Comment by lwsanty** [comment]
@dennwc several points to mention:
- yes I was testing it when submitted the PR
- I was not testing it after all the review fixes done
- we cannot roll it now on staging or elsewhere because it requires installations db's data to be 'fixed'

**Review by dennwc — APPROVED** [review]

**Pull request merged**